Update Shapes app per Marketplace requirements

ctrlaltdavid opened the pull request:
Changes:
- The "SHAPES" tablet/toolbar icon opens a tablet/desktop dialog that has a button that turns the app on/off (shows/hides the "Tools" menu and "Create" palette attached to your non-dominant hand).

Fixes:
- The laser target (without laser beam) appears as expected on the menu and palette when the dominant hand is aimed at it without the trigger pulled.
- When the laser is showing (trigger partly or fully pulled) and the target distance jumps from far to near, or from near to far, the target size does not "flash" big or small when the laser length transitions.
- The icon and text graphics (SVG files) have been reworked to work around an Interface bug where they were not rendered their correct sizes (e.g., text appeared to be displayed at varying font sizes).
- Sometimes lasers drawn in wrong direction and hand spheres not displaying after changing domains.

thoys commented on Marketplace/shapes/html/shapes.html:
```diff
@@ -0,0 +1,79 @@
+        script.onload = function () {
+            ui.onLoad();
+            document.getElementsByTagName("body")[0].setAttribute("onunload", "ui.onUnload()");
+        }
```
missing semi-colon

thoys commented on Marketplace/shapes/shapes.js:
```diff
@@ -1828,30 +1832,75 @@
+        switch (message.command) {
+            case SET_ACTIVE_MESSAGE:
+                switch (message.value) {
+                    case true:
```
dual switching? Only two options, `true` and `false`. What made you choose to have a switch statement over simply having an if-else statement?

ctrlaltdavid replied:
@thoys Defensive programming: `message` is parsed from `data` being a JSON string, and so `message.value` could in theory be `true`, `false`, or `undefined`. In practice, unless there's a code change that produces a bug in the `data` value, it will always be `true` or `false`. Easy to change to an `if()` statement if you still want.

ctrlaltdavid commented:
# USER GUIDE


**Script:** `shapes.js`


## Running

Running the script adds a "SHAPES" tablet/toolbar icon.

The icon is disabled if you start up not connected to a domain or teleport to a domain where you don't have rez or rez-temp permissions. However, once enabled it stays enabled if the domain stops, or if you teleport to a domain that isn't connected. (I.e., same behavior as edit.js's icon.)

Enable or disable editing by clicking the app icon to display a dialog that has a button that toggles the app on and off. The app can only be turned on if you're in a domain where you have rez or rez-temp permissions. The app is automatically turned off in you teleport to a domain where you don't have rez or rez-temp permissions.

When the app is on, the app icon is highlighted and a "Tools" menu and "Create" palette are attached to your non-dominant hand. (Use Settings > Avatar > Dominant Hand to set your dominant hand.)

The UI is hidden if your hand is at or inside an entity but your viewpoint is outside it.


### Grabbing

* This is the default action for each hand with no tool active (i.e., no tool icon on your non-dominant hand).
* Make your hand or blue laser (partially squeezed trigger) intersect an entity. All the entities in that entity's parent/child tree are highlighted in yellow.
* Fully squeeze the trigger to grab the set of entities. The entities stop being highlighted and start translating and rotating 1-1 with your hand translation and rotation.
* You can grab and manipulate two sets of entities simultaneously: one with each hand/laser.
* To delete the grabbed entities, press the grip trigger/button on the controller that's grabbing them. This is an alternative to using the Delete tool.

NOTE: Entities are only highlighted if visible and grabbable if the root of the tree is not locked, and not a zone or light.


### Scale With Hands

* This is the default action when using both hands on the same entity or set of entities.
* Make your second hand/laser intersect an entity in the set you have grabbed with your first hand. All the entities in the set are highlighted in cyan.
* Fully squeeze your second hand's trigger. The entities stop being highlighted and start being scaled and rotated as you move your hands.
* If you let go with one hand, the entities continue to be grabbed by the hand still grabbing them.


## Tools Menu

* To activate a tool, click on its button. An icon representing it is attached to your dominant hand.
* To deactivate a tool, hover the title bar so that it displays "Back" and click it, or click a grip trigger/button on a hand that you're not grabbing entities with. (See also, the Delete tool.)


### Scale With Handles Tool

* Make your hand or blue laser (partially squeezed trigger) intersect an entity. All the entities in that entity's parent/child tree are highlighted in cyan - to indicate "about to scale".
* Fully squeeze the trigger to grab the set of entities. Scale handles are displayed and the entities start translating and rotating 1-1 with your hand translation and rotation.
* Grab a handle with your second hand / laser and move it. The entity is scaled in that axis if a face handle, or as a whole if a corner handle.
* Face handles are not displayed if multiple entities are selected. (Some may be at an angle relative to the root entity which would necessitate a shear transform - not available in Interface - be applied to them when scaling a face of the set.)
* Z-axis handles are not displayed if scaling a text or Web entity.


### Clone Tool

* Make your hand or blue laser (partially squeezed trigger) intersect an entity. 
* Fully squeeze the trigger to clone the set of entities and then start grabbing it.


### Group/Ungroup Tool

* Select one or more sets of entities by clicking on them. Selected entities are highlighted in magenta.
* Green "Group" button: Press this button to group the entities selected in a parent-child relationship. The first entity or entity selected is used as the parent. (Button is enabled only if two or more sets of entities are selected.)
* Red "Ungroup" button: Press this button to ungroup the enttiy set selected. (Button is enabled only if one entity set with two or more entities is selected.)
* "Select in box" dotted rectangle button: When this is toggled on, a bounding box emcompassing your selected items is drawn and all entity sets contained within this box are automatically selected. The orientation of the bounding box is per the orientation of first entity in the first entity set selected.
* "Clear" button: Deselects all entities.
* Grouping and ungrouping is relative to the root entity of the first entity set selected.
* You can group and ungroup more than one level of parent-child relationships.


### Color Tool

* Click on and drag your laser beam in the color circle and slider to change the color in the "current color" patch.
* Click on an entity to apply the color in the "current color" patch to that entity.
* For entity sets, only the intersected entity is highlighted and the color is applied to only that entity.
* Only shape (box, sphere, cylinder, etc.), particle, and polyline entity types can be colored.
* There are several color swatches, empty by default.
* Trigger-click an empty color swatch to set its color to the current color. 
* Grip-click a filled color swatch to clear it to empty. 
* Color swatches persist across Interface runs.
* Click the "Pick" button to change into "pick color" mode, then click on an entity to pick that entity's color as the current color. You can pick colors from both unlocked and locked entities.


The following entity types do not take color:
* Model.
* Web.
* Text.
* Particle.


### Physics Tool

* Adjust the settings in the options panel then click on an entity to apply them to that entity or entity set. 
* Physics settings persist across Interface runs.

The following entity types do not take gravity:
* Web.
* Text.

NOTE: With entity sets, the physics is applied to the root entity and child entities' properties are adjusted as necessary to make physics work within High Fidelity's current limitations w.r.t. physics and entity sets.
NOTE: With entity sets, only the root entity collides with other entities (a current limitation of High Fidelity).


### Delete Tool

* Make your hand or blue laser (partially squeezed trigger) intersect an entity. 
* Fully squeeze the trigger to delete the entity or set of entities.
* Alternatively, you can delete entities without using this tool by grabbing them and clicking the grip trigger/button.


### Undo, Redo

* Press the undo button one or more times to undo your latest edits.
* After undoing some edits, you can redo them by pressing the redo button.
* If you undo some edits then make a new edit, the new edits that you undo are forgotten (i.e., the new edit replaces them in your edit history).
* There's an error beep if there's nothing to undo or redo when you press the button.



## Create Palette

* Click an entity represented in the palette and the entity will be created at your hand.
* If you continue to hold the trigger pressed the entity is automatically grabbed.

ctrlaltdavid commented:
# TEST PLAN

KNOWN ISSUES:
* Line entity types are not supported.


## Running

* General test per User Guide.
* If you're in a domain where you don't have rez or rez-temp permissions and you try to turn the app on, an error beep is emitted when you press the button, a notification message is displayed, and the app UI does not appear on your hand.
* If the app is on and your domain stops or restarts, the app remains on (app UI disappears then reappears) though any current tool is dropped and it returns to the main menu.
* If your permissions are updated so that you no longer have rez or rez-temporary permissions, the UI closes if it is open and the app button is disabled.
* With editing active the following non-editing functions still work:
  * Laser and stylus on tablet UI and "x" button.
  * Grabbing and moving the tablet.
  * Teleporting using the controller teleport button.
* Test basics of operation (e.g., lasers) work when activate the app in desktop mode before entering HMD mode, and when activate app while in HMD mode.


### Grabbing

* General test per User Guide.
* Entity types that can be highlighted and grabbed:
  * Model
  * Shape (box, sphere, cylinder, etc.)
  * Particle effect
  * Text
  * Web
  * PolyLine
* Entity types that cannot be highlighted or grabbed:
  * Zone
  * Light
* Locked entities are highlighted but cannot be grabbed.
* Invisible entities are not highlighted and cannot be grabbed.

KNOWN ISSUES:
* Particle effects cannot be laser-highlighted or laser-grabbed.
* Polylines cannot be laser-highlighted or laser-grabbed.


### Scale With Hands

* General test per User Guide.
* Test scaling entities and groups of entities with center and off-center registration points. Check rotations of children.


## Tools Menu

* General test per User Guide.


### Scale With Handles Tool

* General test per User Guide.
* Test scaling entities and groups of entities with center and off-center registration points. Check rotations of children.
* Test grabbing entities in different positions (e.g., center, one end, etc.).


### Clone Tool

* General test per User Guide.
* Test cloning entities with both hands simultaneously.


### Group/Ungroup Tool

* General test per User Guide.
* Test grouping and ungrouping several levels of groups.
* Test grouping and ungrouping different the entity types able to be highlighted per "Grabbing".


### Color Tool

* General test per User Guide.


### Physics Tool

* General test per User Guide.
* Test grouping and ungrouping entity sets with physics (gravity and collisions) applied.
  * When grouping, the resulting entity set has phycis if the parent selected has physics.
  * When ungrouping, if the entity set has physics then just resulting parent entity or group has physics and the detached children have default (i.e.) no physics.

KNOWN ISSUE: Some shapes other than box and sphere cannot be grabbed with controllerDispatcher.js. FB7333


### Delete Tool

* General test per User Guide.


### Undo, Redo

* General test per User Guide, for all app features.



## Create Palette

* General test per User Guide.
* Test creating entities while a tool is active. The entity should be created without the tool being applied (e.g., entity not colored, etc.).



## Miscellaneous Tests

* You can use the tablet while the app is enabled.
* Test operations while you're inside an entity (e.g., inside a house model): you should be able to grab, create, color, etc., entities the same as if you're not inside an entity. Except that the entity you're inside is not highlighted unless you laser it or one of the entities in its linkset.
* When the app is running and you change your avatar, the UI closes. You can open it again by clicking the app icon.
* The UI is usable (readable) in low-light conditions, i.e., the necessary parts of the UI are bright.
* In your sandbox, unlock the builders grid and check that the UI still displays OK while you're standing on the grid.

YuppyPlays commented:
Testing.

YuppyPlays commented:
**Testing complete:**
[Given the structure of the test plan and user guide, instead of marking everything that works / check out, I will only list issues with detail - I have filed bugs for the issues that are classed as bugs, where as some of these are test plan discrepancies]

**Grabbing:**
**"Entity types that can be highlighted and grabbed"**
**"Entity types that cannot be highlighted or grabbed"**
 - Particle effect is mentioned in both sectons for what is meant to be grabbable and what isn't?.. In any case, this can be highlighted / grabbed but as mentioned in test plan, not with laser but rather up-close grabbing.
 - I am unfamiliar with PolyLines mentioned in Test Plan, I do not see it in Create or Shapes app. Is this the same as for example what Fingerpaint app creates?. Besides this though, all check.

**Color Tool**

**"Only shape (box, sphere, cylinder, etc.), particle, and polyline entity types can be colored."**
 - Still not familiar with PolyLine as mentioned above. Besides this though, all check.


**Miscellaneous Tests**
**"In your sandbox, unlock the builders grid and check that the UI still displays OK while you're standing on the grid."** 
 - I noticed the Locked grids are perfect but unlocked grids have a strange issue where looking up and holding your arm up into your view makes the Shapes UI disappear, this is rectified by locking the grid.
Bug Filed: [16070](https://highfidelity.fogbugz.com/f/cases/16070/Hifi-Content-Shapes-App-152-Builders-grid-unlock-issue)

**Other:**
 - I experienced quite a few crashes whilst using the Shapes app though I am uncertain if the crashes were directly connected with the app itself, they seemed to happen whenever I used advanced movement specifically (example, whilst testing, I'd use Left on the left-vive-controllers thumbpad to move left, insta-crash) I cannot seem to find an exact pin-point for it other than having shapes app and then using advanced movement.
Bug Filed: [16071](https://highfidelity.fogbugz.com/f/cases/16071/Hifi-Content-Shapes-App-152-Crashing-when-using-Advanced-Movement)


This is a **fail** for now I'd have to say mainly for the two bugs mentioned, I am happy to quickly go through the PolyLines to ensure functionality with Shapes App if someone can confirm my query about them mentioned above.

(PS: Love this app already, Easy and intuitive just like Google Blocks)

ctrlaltdavid commented:
@YuppyPlays PolyLines can be created using the Paint app.

YuppyPlays commented:
Re-tested with Fingerpaint app for the PolyLines where required- all work perfectly with close-grab interactions.

Although, I did notice a bug whilst doing this regarding the grab-laser, for some reason (this has happened to me twice now though the initial time I assumed it was just an interface issue) the laser seems to go off into the sky regardless of where I aim, which is confusing as it only happens when Shapes app is open.
Bug Filed: [16084](https://highfidelity.fogbugz.com/f/cases/16084/Hifi-Content-Shapes-App-152-Far-grab-laser-issue)

ctrlaltdavid commented:
*ParticleEffect entities*: These shouldn't be able to be grabbed or otherwise operated upon. I've updated the User Guide, Test Plan, and Marketplace description.

*Builder's Grid*: Cannot reproduce. Hand likely intersected an "invisible" ParticleEffect entity (version tested operated on these) and so changed to "grabbing" mode with yellow sphere on hand and the entity's bounding box edges rendered in yellow. https://highfidelity.fogbugz.com/f/cases/16070/Hifi-Content-Shapes-App-152-Builders-grid-unlock-issue

*Crashes while strafing*: Build 8468 fixes crashes while strafing. https://highfidelity.fogbugz.com/f/cases/16071/Hifi-Content-Shapes-App-152-Crashing-when-using-Advanced-Movement

*Laser in sky*: I've not managed to reproduce this. https://highfidelity.fogbugz.com/f/cases/16084/Hifi-Content-Shapes-App-152-Far-grab-laser-issue Though I did get a case of two laser beams if you turn the Shapes app on while far-grabbing the entity. However, this appears benign: the entity can still be moved, and both laser beams go off when you release the trigger. https://highfidelity.manuscript.com/f/cases/16093/Shapes-app-Two-laser-beams-if-turn-app-on-while-far-grabbing-an-entity

I've updated the User Guild and Test Paln but there are no code changes.

Testing: Please retest the draft Marketplace submission with the current public release of High Fidelity

YuppyPlays commented:
Testing.

YuppyPlays commented:
**Testing complete:**

[Using interface build 8413 (Current RC) & latest Shapes.js provided by David]

The crashing whilst strafing didn't happen once this time round.

**Everything is a pass except for the following experienced issues:**
 - Builders grid issue - did a check to make sure there were no intersecting particleeffect hitboxes.
 - Laser issue / Laser in sky (still persistent unfortunately)
 - Menu flashing/loading in (this did not occur in the previous test conditions, I am unsure if this counts as a bug or just somewhat normal engine behaviour of sorts?.. Need feedback)

I quickly put together a video as discussed to show all mentioned issues in better detail as they can be a bit difficult to describe, located [here](https://youtu.be/cO2wycgIHEI)

ctrlaltdavid commented:
- Builder's grid issue: There's something amiss with either the builders grid model or Entities.findRayIntersection(): the ray is intersecting with the bounding box in addition to the model's surface.
- Laser issue: This also affected the hand spheres. Fixed.
- UI flashing / slow loading: This is just a side-effect of running the app from the Marketplace staging area. (All files are downloaded afresh each time they're accessed rather than the cache being used; the files are read into cache at program start so in production they'll appear promptly.)

YuppyPlays commented:
Testing complete.

Using the fix provided by David (http://mpassets.highfidelity.com/fd00bebd-0180-4083-9123-625cc996d0a2-v1/shapes.js) - I can confirm in build 8413, the issue did not occur for me, repeatedly went back and fourth from Zaru and Help domains back to my own sandbox to see if issue would happen in the same fashion as before, as said, did not :)
